### PR TITLE
Anuragd/multi_gpu_device Fixed multi-gpu scenario with Python set_device API support

### LINKED ISSUE #356: 🐛 [Bug] An error occurs in CompileGraph when gpu_id == 1
When I tried to Complie on the second GPU in a multi-GPU environment, an error occurred. The code sample used is as follows.

```cpp
void load(const std::string& model_path, int64_t gpu_id, int64_t opt_batch_size) {
    torch::jit::Module module = torch::jit::load(model_path);
    torch::Device device = (torch::cuda::is_available() ? torch::Device(torch::kCUDA, gpu_id) : torch::Device(torch::kCPU));
    module.to(device, torch::kHalf);
    module.eval();

    std::vector<int64_t> in_opt = { opt_batch_size, INPUT_CHANNEL_NUM, BOARD_WIDTH, BOARD_WIDTH };

    trtorch::CompileSpec::InputRange range(in_opt);
    trtorch::CompileSpec info({ range });
    info.op_precision = torch::kHalf;
    info.device.gpu_id = gpu_id;
    module = trtorch::CompileGraph(module, info);
}
```

#### Error1
I called this function with gpu_id = 1. I got the following error:

```
terminate called after throwing an instance of 'trtorch::Error'
  what():  [enforce fail at core/conversion/conversionctx/ConversionCtx.cpp:107] Expected cudaSetDevice(settings.device.gpu_id) to be true but got false
Unable to set gpu id: 1
```

I think this line is the cause.
https://github.com/NVIDIA/TRTorch/blob/1d4b967a28e36beee048703f5645ee6fcc95793d/core/conversion/conversionctx/ConversionCtx.cpp#L112

`cudaSetDevice` returns `cudaSuccess` (= 0) on success. However, `TRTORCH_CHECK` judges success or failure as a Boolean type.

I fixed it as follows and rebuilt it so that this error disappeared.

```diff
diff --git a/core/conversion/conversionctx/ConversionCtx.cpp b/core/conversion/conversionctx/ConversionCtx.cpp
index ff23692..bc5bf68 100644
--- a/core/conversion/conversionctx/ConversionCtx.cpp
+++ b/core/conversion/conversionctx/ConversionCtx.cpp
@@ -109,7 +109,7 @@ ConversionCtx::ConversionCtx(BuilderSettings build_settings)
   cfg->setEngineCapability(settings.capability);
 
   if (settings.device.gpu_id) {
-    TRTORCH_CHECK(cudaSetDevice(settings.device.gpu_id), "Unable to set gpu id: " << settings.device.gpu_id);
+    TRTORCH_CHECK(cudaSetDevice(settings.device.gpu_id) == cudaSuccess, "Unable to set gpu id: " << settings.device.gpu_id);
   }
 
   if (settings.device.device_type == nvinfer1::DeviceType::kDLA) {
```

You may also use `set_device`.

https://github.com/NVIDIA/TRTorch/blob/1d4b967a28e36beee048703f5645ee6fcc95793d/core/compiler.cpp#L176-L178

#### Error2
After making the above fix, I get the following error:

```
ERROR: [TRTorch Conversion Context] - Builder was created on device different than current device.
```

I changed `cudaSetDevice` to do it at the beginning of the function and it worked fine.

```diff
diff --git a/core/conversion/conversionctx/ConversionCtx.cpp b/core/conversion/conversionctx/ConversionCtx.cpp
index ff23692..09a419c 100644
--- a/core/conversion/conversionctx/ConversionCtx.cpp
+++ b/core/conversion/conversionctx/ConversionCtx.cpp
@@ -47,6 +47,10 @@ ConversionCtx::ConversionCtx(BuilderSettings build_settings)
           util::logging::get_logger().get_reportable_severity(),
           util::logging::get_logger().get_is_colored_output_on()) {
   // TODO: Support FP16 and FP32 from JIT information
+  if (settings.device.gpu_id) {
+    TRTORCH_CHECK(cudaSetDevice(settings.device.gpu_id) == cudaSuccess, "Unable to set gpu id: " << settings.device.gpu_id);
+  }
+
   builder = nvinfer1::createInferBuilder(logger);
   net = builder->createNetworkV2(1U << static_cast<uint32_t>(nvinfer1::NetworkDefinitionCreationFlag::kEXPLICIT_BATCH));
 
@@ -108,10 +112,6 @@ ConversionCtx::ConversionCtx(BuilderSettings build_settings)
   cfg->setDefaultDeviceType(settings.device.device_type);
   cfg->setEngineCapability(settings.capability);
 
-  if (settings.device.gpu_id) {
-    TRTORCH_CHECK(cudaSetDevice(settings.device.gpu_id), "Unable to set gpu id: " << settings.device.gpu_id);
-  }
-
   if (settings.device.device_type == nvinfer1::DeviceType::kDLA) {
     auto nbDLACores = builder->getNbDLACores();
     TRTORCH_CHECK(
```

It's working, but I'm not sure if this is a good fix as there may be other side effects as well.
I would appreciate it if you could respond appropriately.

### PR DESCRIPTION
**Pull request by andi4191**
# Description

Fix for multi-gpu configuration where the compilation of graph fails.
Root cause: Implicit typecasting of cudaSucess as boolean. Incorrect cuda context before setting the device ID (ConversionCtx)

Fixes #356 

## Type of change

Please delete options that are not relevant and/or add your own.
- Bug fix (non-breaking)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [x] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by peri044 — APPROVED** [review]
LGTM. Multi_gpu_test and other tests ran fine on my machine.

File: docs/_notebooks/Resnet50-example.html [hunk 1]
```diff
@@ -1932,4 +1932,4 @@ <h3 id="What's-next">
    app.initialize({version: "1.0.4", url: {base: ".."}})
```
Comment (narendasan) [inline docs/_notebooks/Resnet50-example.html:1932]: Remove changes to the rendered docs

File: docsrc/tutorials/getting_started.rst [hunk 2]
```diff
@@ -201,6 +201,7 @@ to load in a deployment application. In order to load a TensorRT/TorchScript mod
 
     ...
 
+    trtorch.set_device(0) # Set CUDA device (in case of multi-gpu configuration). Defaults to 0
```
Comment (narendasan) [inline docsrc/tutorials/getting_started.rst:204]: I dont think that its hugely important for this to be in the getting started docs. Its more imporant to cover the semantics of selected device in the compile spec

File: tests/py/test_api_multi_gpu.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
+import unittest
+import trtorch
+import torch
+import torchvision.models as models
+
+from model_test_case import ModelTestCase
+
+gpu_id = 1
+class TestCompile(ModelTestCase):
```
Comment (narendasan) [inline tests/py/test_api_multi_gpu.py:9]: Why are we duplicating tests for logging and op support?

Comment (narendasan) [inline tests/py/test_api_multi_gpu.py:9]: Make this test name more descriptive

File: tests/py/test_api_multi_gpu.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
+import unittest
+import trtorch
+import torch
+import torchvision.models as models
+
+from model_test_case import ModelTestCase
+
+gpu_id = 1
```
Comment (narendasan) [inline tests/py/test_api_multi_gpu.py:8]: theres no point in making this redundant declaration we can just set self.gpu_id to 1.

File: tests/py/test_api_multi_gpu.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+class TestCompile(ModelTestCase):
+
+    def setUp(self):
+        self.gpu_id = gpu_id
```
Comment (narendasan) [inline tests/py/test_api_multi_gpu.py:12]: Check during setup if there is more than one GPU (you can probably do this through pytorch apis). If there is not then error out with a message saying this test is not valid on this platform.

File: tests/py/test_api_multi_gpu.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                "disable_tf32": False
+            }
+        }
+
+        trt_mod = trtorch.compile(self.scripted_model, compile_spec)
+        same = (trt_mod(self.input) - self.scripted_model(self.input)).abs().max()
+        self.assertTrue(same < 2e-3)
+
+
+class TestCheckMethodOpSupport(unittest.TestCase):
```
Comment (narendasan) [inline tests/py/test_api_multi_gpu.py:50]: Remove redundant test

File: tests/py/test_api_multi_gpu.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    def setUp(self):
+        module = models.alexnet(pretrained=True).eval().to("cuda")
+        self.module = torch.jit.trace(module, torch.ones((1, 3, 224, 224)).to("cuda"))
+
+    def test_check_support(self):
+        self.assertTrue(trtorch.check_method_op_support(self.module, "forward"))
+
+
+class TestLoggingAPIs(unittest.TestCase):
+
```
Comment (narendasan) [inline tests/py/test_api_multi_gpu.py:61]: remove redundant test

**Review by narendasan — CHANGES_REQUESTED** [review]

File: docs/_notebooks/Resnet50-example.html (hunk 1 above)
Comment (andi4191) [reply docs/_notebooks/Resnet50-example.html:1932]: Fixed

File: docsrc/tutorials/getting_started.rst (hunk 2 above)
Comment (andi4191) [reply docsrc/tutorials/getting_started.rst:204]: Added it for reference as to how to use it. Removed it anyways.

File: tests/py/test_api_multi_gpu.py (hunk 3 above)
Comment (andi4191) [reply tests/py/test_api_multi_gpu.py:9]: Intial plan was to run either test_api.py or test_api_multi_gpu.py via Bazel BUILD file. Removed it now

Comment (andi4191) [reply tests/py/test_api_multi_gpu.py:9]: Done

File: tests/py/test_api_multi_gpu.py (hunk 4 above)
Comment (andi4191) [reply tests/py/test_api_multi_gpu.py:8]: Single point of configuration. Using same variable at all the places in the test script

File: tests/py/test_api_multi_gpu.py (hunk 5 above)
Comment (andi4191) [reply tests/py/test_api_multi_gpu.py:12]: Dropped

Comment (andi4191) [reply tests/py/test_api_multi_gpu.py:12]: Added the device_count test

File: tests/py/test_api_multi_gpu.py (hunk 6 above)
Comment (andi4191) [reply tests/py/test_api_multi_gpu.py:50]: done

File: tests/py/test_api_multi_gpu.py (hunk 7 above)
Comment (andi4191) [reply tests/py/test_api_multi_gpu.py:61]: done

**Comment by andi4191** [comment]
Due to multiple conflicts with the docs/ I have moved the PR as a new PR  (with address to review comments)
Please refer https://github.com/NVIDIA/TRTorch/pull/371

Closing this PR.

**Pull request closed without merging**